revlist/revlist.go

```diff
@@ -0,0 +1,250 @@
+	for _, c := range l {
+		buf.WriteString(c.Hash.String()[:8])
+		buf.WriteString("\n")
+	}
+	return buf.String()
+}
+
+// New returns a Revs pointer for the
+// file at "path", from commit "commit" backwards in time.
+// The commits are stored in arbitrary order.
+// It stops searching a branch for a file upon reaching the commit
+// were it was created.
+// Moves and copies are not currently supported.
+// Cherry-picks are not detected and therefore are added to the list
+// (see git path-id for hints on how to fix this).
+// This function implements is equivalent to running go-rev-Revs.
+func New(repo *git.Repository, commit *git.Commit, path string) (Revs, error) {
+	result := make(Revs, 0)
+	seen := make(map[core.Hash]struct{}, 0)
+	err := walkGraph(&result, &seen, repo, commit, path)
+	if err != nil {
+		return nil, err
+	}
+	sort.Sort(result)
+	result = removeComp(path, result, equivalent) // for merges of identical cherry-picks
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Recursive traversal of the commit graph, generating a linear history
+// of the path.
+func walkGraph(result *Revs, seen *map[core.Hash]struct{}, repo *git.Repository, current *git.Commit, path string) error {
+	// check and update seen
+	if _, ok := (*seen)[current.Hash]; ok {
+		return nil
+	}
+	(*seen)[current.Hash] = struct{}{}
+
+	// if the path is not in the current commit, stop searching.
+	if _, found := git.FindFile(path, current); !found {
+		return nil
+	}
+
+	parents := parentsContainingPath(path, current)
+
+	switch len(parents) {
+	// if the path is not found in any of its parents, the path was
+	// created by this commit; we must add it to the revisions list and
+	// stop searching. This includes the case when current is the
+	// initial commit.
+	case 0:
+		//fmt.Println(current.Hash.String(), ": case 0")
+		*result = append(*result, current)
+		return nil
+	case 1: // only one parent contains the path
+		// if the file contents has change, add the current commit
+		different, err := differentContents(path, current, parents)
+		if err != nil {
+			return err
+		}
+		if len(different) == 1 {
+			//fmt.Println(current.Hash.String(), ": case 1")
+			*result = append(*result, current)
+		}
+		// in any case, walk the parent
+		return walkGraph(result, seen, repo, parents[0], path)
+	default: // more than one parent contains the path
+		// TODO: detect merges that had a conflict, because they must be
+		// included in the result here.
+		for _, p := range parents {
+			err := walkGraph(result, seen, repo, p, path)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// TODO: benchmark this making git.Commit.parent public instead of using
+// an iterator
+func parentsContainingPath(path string, c *git.Commit) []*git.Commit {
+	var result []*git.Commit
+	iter := c.Parents()
+	for {
+		parent, err := iter.Next()
+		if err != nil {
+			if err == io.EOF {
+				return result
+			}
+			panic("unreachable")
```

Comment:
I'd say yer, we need them.

Without the panic, If iter.Next() changes in the future, and start returning other errors, they will be invisible to our code.

This way, the problem can be easily identified and fixed.